Index: control
===================================================================
--- control	(Revision 381)
+++ control	(Arbeitskopie)
@@ -4,7 +4,7 @@
 Maintainer: Debian GGZ Maintainers <pkg-ggz-maintainers@lists.alioth.debian.org>
 Uploaders: Josef Spillner <josef@ggzgamingzone.org>, Peter Eisentraut <petere@debian.org>
 Standards-Version: 3.7.3
-Build-Depends: cdbs, debhelper (>= 5), libgnutls-dev
+Build-Depends: cdbs, debhelper (>= 5)
 Vcs-Svn: svn://svn.debian.org/pkg-ggz/trunk/libggz/
 Vcs-Browser: http://svn.debian.org/wsvn/pkg-ggz/trunk/libggz/
 Homepage: http://www.ggzgamingzone.org/
@@ -26,7 +26,7 @@
 Package: libggz-dev
 Architecture: any
 Section: libdevel
-Depends: libggz2 (= ${binary:Version}), libgnutls-dev
+Depends: libggz2 (= ${binary:Version})
 Description: GGZ Gaming Zone: common utilities library - development files
  The libggz library handles network, configuration file and data
  structure management functions as well as several convenience wrappers
Index: changelog
===================================================================
--- changelog	(Revision 381)
+++ changelog	(Arbeitskopie)
@@ -1,3 +1,11 @@
+libggz (0.0.14.1-1maemo0) diablo; urgency=low
+
+  [ Josef Spillner ]
+  * Port to the Maemo platform, release Diablo (4.1.1).
+  * Disabled gnutls for that matter.
+
+ -- Josef Spillner <josef@ggzgamingzone.org>  Sat, 25 Oct 2008 17:42:25 +0200
+
 libggz (0.0.14.1-1) unstable; urgency=low
 
   [ Peter Eisentraut ]
Index: rules
===================================================================
--- rules	(Revision 381)
+++ rules	(Arbeitskopie)
@@ -3,5 +3,5 @@
 include /usr/share/cdbs/1/rules/debhelper.mk
 include /usr/share/cdbs/1/class/autotools.mk
 
-DEB_CONFIGURE_EXTRA_FLAGS = --with-tls --with-gcrypt --disable-debug
+DEB_CONFIGURE_EXTRA_FLAGS = --disable-debug
 DEB_DH_MAKESHLIBS_ARGS_ALL = -V
