diff -Nur src.20020114_19-13-00/ggz_common.c src/ggz_common.c
--- src.20020114_19-13-00/ggz_common.c	Mon Jan 14 19:13:17 2002
+++ src/ggz_common.c	Mon Jan 14 19:12:33 2002
@@ -41,6 +41,8 @@
 		return "reserved";
 	case GGZ_SEAT_PLAYER:
 		return "player";
+	case GGZ_SEAT_SPECTATOR:
+		return "spectator";
 	default:
 		return "none";
 	}
@@ -59,6 +61,8 @@
 		return GGZ_SEAT_RESERVED;
 	else if (!strcmp(type_str, "player"))
 		return GGZ_SEAT_PLAYER;
+	else if (!strcmp(type_str, "spectator"))
+		return GGZ_SEAT_SPECTATOR;
 	else
 		return GGZ_SEAT_NONE;
 }
diff -Nur src.20020114_19-13-00/ggz_common.h src/ggz_common.h
--- src.20020114_19-13-00/ggz_common.h	Mon Jan 14 19:13:17 2002
+++ src/ggz_common.h	Mon Jan 14 19:11:42 2002
@@ -44,7 +44,8 @@
 	GGZ_SEAT_BOT = -2,	   /**< The seat has a bot (AI) in it. */
 	GGZ_SEAT_RESERVED = -3,	   /**< The seat is reserved for a player. */
 	GGZ_SEAT_NONE = -4,	   /**< This seat does not exist. */
-	GGZ_SEAT_PLAYER = -5	   /**< The seat has a regular player in it. */
+	GGZ_SEAT_PLAYER = -5,	   /**< The seat has a regular player in it. */
+	GGZ_SEAT_SPECTATOR = -6 /**< A spectator watches the game on this seat. */
 } GGZSeatType;
 
 /** @brief Get a string identifier for the GGZSeatType.
