-/configure.in/1.63/Sat Feb 23 02:16:51 2002//
diff -Nur ggzd/configure.in ggzd.tls/configure.in
--- ggzd/configure.in	Sat Feb 23 03:16:51 2002
+++ ggzd.tls/configure.in	Sat Feb 23 03:14:22 2002
@@ -97,7 +97,7 @@
 dnl ==============
 
 AC_GGZ_LIBGGZ
-LDADD="$LDADD $LIB_GGZ"
+LDADD="$LDADD $LIB_GGZ -lggztls"
 
 
 dnl Check for expat
diff -Nur ggzd/ggzd/control.c ggzd.tls/ggzd/control.c
--- ggzd/ggzd/control.c	Sat Feb 23 03:16:33 2002
+++ ggzd.tls/ggzd/control.c	Sat Feb 23 03:14:35 2002
@@ -177,6 +177,7 @@
 	dbg_msg(GGZ_DBG_CONFIGURATION, "Conf file: %s", opt.local_conf);
 	dbg_msg(GGZ_DBG_CONFIGURATION, "Log level: %0X", log_info.log_types);
 	dbg_msg(GGZ_DBG_CONFIGURATION, "Main Port: %d", opt.main_port);
+	dbg_msg(GGZ_DBG_CONFIGURATION, "Encryption in use: %d", opt.tls_use);
 	
 	init_dirs();
 	init_data();
diff -Nur ggzd/ggzd/datatypes.h ggzd.tls/ggzd/datatypes.h
--- ggzd/ggzd/datatypes.h	Sat Feb 23 03:16:33 2002
+++ ggzd.tls/ggzd/datatypes.h	Sat Feb 23 03:14:35 2002
@@ -54,6 +54,10 @@
 	int perform_lookups;
 	int ping_freq;
 	int lag_class[4];
+	int tls_use;
+	char *tls_key;
+	char *tls_cert;
+	char *tls_password;
 } Options;
 
 
diff -Nur ggzd/ggzd/ggzd.conf.in ggzd.tls/ggzd/ggzd.conf.in
--- ggzd/ggzd/ggzd.conf.in	Sat Feb 23 03:16:47 2002
+++ ggzd.tls/ggzd/ggzd.conf.in	Sat Feb 23 03:14:35 2002
@@ -42,6 +42,12 @@
 #IPBanList = 208.137.159.0/255.255.255.0
 #  (bans all addresses from 208.137.159.0 - 208.137.159.255)
 
+# Transport Layer Security options
+#EncryptionUse = 1
+#EncryptionPassword = ggzrocksmore
+#EncryptionKey = /usr/local/ssl/certs/openssl-example.pem
+#EncryptionCert = /usr/local/ssl/certs/openssl-example.pem
+
 ############################################################################
 # Logging Options
 ############################################################################
diff -Nur ggzd/ggzd/net.c ggzd.tls/ggzd/net.c
--- ggzd/ggzd/net.c	Sat Feb 23 03:16:34 2002
+++ ggzd.tls/ggzd/net.c	Sat Feb 23 03:14:35 2002
@@ -39,6 +39,7 @@
 #include <errno.h>
 #include <expat.h>
 #include <ggz.h>
+#include <ggz_tls.h>
 #include <string.h>
 #include <sys/poll.h>
 #include <sys/types.h>
@@ -123,6 +124,7 @@
 static void _net_handle_motd(GGZNetIO *net, GGZXMLElement *element);
 static void _net_handle_data(GGZNetIO *net, GGZXMLElement *element);
 static void _net_handle_pong(GGZNetIO *net, GGZXMLElement *element);
+static void _net_handle_tls_start(GGZNetIO *net, GGZXMLElement *element);
 
 /* Utility functions */
 static int safe_atoi(char *string);
@@ -251,14 +253,17 @@
 
 /* net_send_XXX() functions for sending messages to the client */
 
-int net_send_serverid(GGZNetIO *net, char *srv_name)
+int net_send_serverid(GGZNetIO *net, char *srv_name, int use_tls)
 {
 	char *xml_srv_name;
 
 	xml_srv_name = ggz_xml_escape(srv_name);
 
+	ggz_tls_disable_fd(net->fd);
+
 	_net_send_line(net, "<SESSION>");
-	_net_send_line(net, "\t<SERVER ID='GGZ-%s' NAME='%s' VERSION='%d' STATUS='%s'>", VERSION, xml_srv_name, GGZ_CS_PROTO_VERSION, "ok");
+	_net_send_line(net, "\t<SERVER ID='GGZ-%s' NAME='%s' VERSION='%d' STATUS='%s' TLS_SUPPORT='%s'>", VERSION, xml_srv_name, GGZ_CS_PROTO_VERSION, "ok",
+		((use_tls && ggz_tls_support_query()) ? "yes" : "no"));
 	_net_send_line(net, "\t\t<OPTIONS CHATLEN='%d'/>", MAX_CHAT_LEN);
 	_net_send_line(net, "\t</SERVER>");
 
@@ -694,7 +699,7 @@
 	buf[0] = '\0';
 	for (i = 0; i < size; i++) {
 		sprintf(buf, "%d ", data[i]);
-		write(net->fd, buf, strlen(buf));
+		ggz_tls_write(net->fd, buf, strlen(buf));
 	}
 	_net_send_string(net, "]]></DATA>");
 
@@ -753,7 +758,7 @@
 		err_sys_exit("Couldn't allocate buffer");
 
 	/* Read in data from socket */
-	if ( (len = read(net->fd, buf, BUFFSIZE)) < 0) {
+	if ( (len = ggz_tls_read(net->fd, buf, BUFFSIZE)) < 0) {
 		
 		/* If it's a non-blocking socket and there isn't data,
                    we get EAGAIN.  It's safe to just return */
@@ -908,6 +913,8 @@
 		process_func = _net_handle_data;
 	else if (strcmp(tag, "PONG") == 0)
 		process_func = _net_handle_pong;
+	else if (strcmp(tag, "TLS_START") == 0)
+		process_func = _net_handle_tls_start;
 	else
 		process_func = NULL;
 	
@@ -1515,6 +1522,12 @@
 }
 
 
+/* Function for <TLS_START> tag */
+static void _net_handle_tls_start(GGZNetIO *net, GGZXMLElement *data)
+{
+	ggz_tls_enable_fd(net->fd, GGZ_TLS_SERVER, GGZ_TLS_VERIFY_NONE);
+}
+
 /************ Utility/Convenience functions *******************/
 
 static int safe_atoi(char *string)
@@ -1627,7 +1640,7 @@
 	vsprintf(buf, line, ap);
 	va_end(ap);
 	strcat(buf, "\n");
-	return write(net->fd, buf, strlen(buf));
+	return ggz_tls_write(net->fd, buf, strlen(buf));
 }
 
 
@@ -1639,5 +1652,5 @@
 	va_start(ap, fmt);
 	vsprintf(buf, fmt, ap);
 	va_end(ap);
-	return write(net->fd, buf, strlen(buf));
+	return ggz_tls_write(net->fd, buf, strlen(buf));
 }
diff -Nur ggzd/ggzd/net.h ggzd.tls/ggzd/net.h
--- ggzd/ggzd/net.h	Sat Feb 23 03:16:34 2002
+++ ggzd.tls/ggzd/net.h	Sat Feb 23 03:14:35 2002
@@ -54,7 +54,7 @@
 void net_free(GGZNetIO* net);
 
 /* Functions to send data to the client */
-int net_send_serverid(GGZNetIO *net, char *srv_name);
+int net_send_serverid(GGZNetIO *net, char *srv_name, int use_tls);
 int net_send_server_full(GGZNetIO *net, char *srv_name);
 int net_send_login(GGZNetIO *net, GGZLoginType type, char status, char *password);
 int net_send_motd(GGZNetIO *net);
diff -Nur ggzd/ggzd/parse_opt.c ggzd.tls/ggzd/parse_opt.c
--- ggzd/ggzd/parse_opt.c	Sat Feb 23 03:16:34 2002
+++ ggzd.tls/ggzd/parse_opt.c	Sat Feb 23 03:14:35 2002
@@ -266,6 +266,12 @@
 	ggz_conf_read_list(ch, "General", "IPBanList", &t_count, &t_list);
 	player_set_ip_ban_list(t_count, t_list);
 
+	/* Enctyption in [General] */
+	opt.tls_use = ggz_conf_read_int(ch, "General", "EncryptionUse", 0);
+	opt.tls_password = ggz_conf_read_string(ch, "General", "EncryptionPassword", NULL);
+	opt.tls_cert = ggz_conf_read_string(ch, "General", "EncryptionCert", NULL);
+	opt.tls_key = ggz_conf_read_string(ch, "General", "EncryptionKey", NULL);
+
 	/* [Directories] */
 	opt.game_dir = ggz_conf_read_string(ch, "Directories", "GameDir", NULL);
 	opt.conf_dir = ggz_conf_read_string(ch, "Directories", "ConfDir", NULL);
diff -Nur ggzd/ggzd/players.c ggzd.tls/ggzd/players.c
--- ggzd/ggzd/players.c	Sat Feb 23 03:16:34 2002
+++ ggzd.tls/ggzd/players.c	Sat Feb 23 03:14:35 2002
@@ -202,7 +202,7 @@
 	}
 
 	/* Send server ID */
-	if (net_send_serverid(player->net, opt.server_name) < 0) {
+	if (net_send_serverid(player->net, opt.server_name, opt.tls_use) < 0) {
 		free(player);
 		pthread_exit(NULL);
 	}
