Index: ggzcore/ggzcore.h
===================================================================
RCS file: /cvsroot/ggz/ggz/src/ggzcore/ggzcore.h,v
retrieving revision 1.86
diff -u -3 -p -r1.86 ggzcore.h
--- ggzcore/ggzcore.h	5 May 2002 21:13:29 -0000	1.86
+++ ggzcore/ggzcore.h	20 May 2002 17:21:57 -0000
@@ -584,6 +584,7 @@ int ggzcore_room_chat(GGZRoom *room,
 
 int ggzcore_room_launch_table(GGZRoom *room, GGZTable *table);
 int ggzcore_room_join_table(GGZRoom *room, const unsigned int num);
+int ggzcore_room_join_table_spectator(GGZRoom *room, const unsigned int num);
 int ggzcore_room_leave_table(GGZRoom *room, int force);
 
 
Index: ggzcore/net.h
===================================================================
RCS file: /cvsroot/ggz/ggz/src/ggzcore/net.h,v
retrieving revision 1.29
diff -u -3 -p -r1.29 net.h
--- ggzcore/net.h	5 May 2002 21:13:29 -0000	1.29
+++ ggzcore/net.h	20 May 2002 17:21:57 -0000
@@ -75,6 +75,7 @@ int _ggzcore_net_send_chat(struct _GGZNe
 
 int _ggzcore_net_send_table_launch(struct _GGZNet *net, struct _GGZTable *table);
 int _ggzcore_net_send_table_join(struct _GGZNet *net, const unsigned int num);
+int _ggzcore_net_send_table_join_spectator(struct _GGZNet *net, const unsigned int num);
 int _ggzcore_net_send_table_leave(struct _GGZNet *net, int force);
 int _ggzcore_net_send_table_seat_update(struct _GGZNet *net, struct _GGZTable *table, struct _GGZSeat *seat);
 int _ggzcore_net_send_table_desc_update(struct _GGZNet *net, struct _GGZTable *table, const char *desc);
Index: ggzcore/netxml.c
===================================================================
RCS file: /cvsroot/ggz/ggz/src/ggzcore/netxml.c,v
retrieving revision 1.29
diff -u -3 -p -r1.29 netxml.c
--- ggzcore/netxml.c	5 May 2002 21:13:29 -0000	1.29
+++ ggzcore/netxml.c	20 May 2002 17:21:58 -0000
@@ -513,6 +513,16 @@ int _ggzcore_net_send_table_join(struct 
 	return status;
 }
 
+int _ggzcore_net_send_table_join_spectator(struct _GGZNet *net, const unsigned int num)
+{
+	int status = 0;
+
+	ggz_debug("GGZCORE:NET", "Sending table join-as-spectator request");
+	_ggzcore_net_send_line(net, "<JOINSPECTATOR TABLE='%d'/>", num);
+
+	return status;
+}
+
 
 int _ggzcore_net_send_table_leave(struct _GGZNet *net, int force)
 {
@@ -859,6 +869,8 @@ static void _ggzcore_net_handle_result(G
 		else if  (strcmp(action, "launch") == 0)
 			_ggzcore_room_set_table_launch_status(room, code);
 		else if  (strcmp(action, "join") == 0)
+			_ggzcore_room_set_table_join_status(room, code);
+		else if  (strcmp(action, "joinspectator") == 0)
 			_ggzcore_room_set_table_join_status(room, code);
 		else if  (strcmp(action, "leave") == 0)
 			_ggzcore_room_set_table_leave_status(room, code);
Index: ggzcore/room.c
===================================================================
RCS file: /cvsroot/ggz/ggz/src/ggzcore/room.c,v
retrieving revision 1.48
diff -u -3 -p -r1.48 room.c
--- ggzcore/room.c	5 May 2002 21:13:29 -0000	1.48
+++ ggzcore/room.c	20 May 2002 17:21:58 -0000
@@ -285,6 +285,15 @@ int ggzcore_room_join_table(GGZRoom *roo
 }
 
 
+int ggzcore_room_join_table_spectator(GGZRoom *room, const unsigned int num)
+{
+	if (room && room->server)
+		return _ggzcore_room_join_table_spectator(room, num);
+	else
+		return -1;
+}
+
+
 int ggzcore_room_leave_table(GGZRoom *room, int force)
 {
 	if (room && room->server)
@@ -882,6 +891,28 @@ int _ggzcore_room_join_table(struct _GGZ
 
 	net = _ggzcore_server_get_net(room->server);
 	status = _ggzcore_net_send_table_join(net, num);
+
+	if (status == 0)
+		_ggzcore_server_set_table_joining(room->server);
+	
+	return status;
+}
+
+
+int _ggzcore_room_join_table_spectator(struct _GGZRoom *room, const unsigned int num)
+{
+	int status;
+	struct _GGZNet *net;
+
+	/* Make sure we're actually in a room (FIXME: should probably
+           make sure we're in *this* room) and not already playing a
+           game */
+	if (_ggzcore_server_get_state(room->server) != GGZ_STATE_IN_ROOM)
+		return -1;
+
+
+	net = _ggzcore_server_get_net(room->server);
+	status = _ggzcore_net_send_table_join_spectator(net, num);
 
 	if (status == 0)
 		_ggzcore_server_set_table_joining(room->server);
Index: ggzcore/room.h
===================================================================
RCS file: /cvsroot/ggz/ggz/src/ggzcore/room.h,v
retrieving revision 1.23
diff -u -3 -p -r1.23 room.h
--- ggzcore/room.h	21 Mar 2002 18:14:52 -0000	1.23
+++ ggzcore/room.h	20 May 2002 17:21:58 -0000
@@ -134,6 +134,7 @@ int _ggzcore_room_chat(struct _GGZRoom *
 
 int _ggzcore_room_launch_table(struct _GGZRoom *room, struct _GGZTable *table);
 int _ggzcore_room_join_table(struct _GGZRoom *room, const unsigned int num);
+int _ggzcore_room_join_table_spectator(struct _GGZRoom *room, const unsigned int num);
 int _ggzcore_room_leave_table(struct _GGZRoom *room, int force);
 
 int  _ggzcore_room_send_game_data(struct _GGZRoom *room, char *buffer);
