Index: java/src/ggz/client/core/Net.java
===================================================================
--- java/src/ggz/client/core/Net.java	(revision 9253)
+++ java/src/ggz/client/core/Net.java	(working copy)
@@ -529,6 +529,9 @@
             sendStartElement("ADMIN", attributes);
             sendTextElement("REASON", reason);
             sendEndElement("ADMIN");
+        } else if (type == AdminType.GGZ_ADMIN_BAN) {
+            attributes.addAttribute("", "", "ACTION", "CDATA", "ban");
+            sendEmptyElement("ADMIN", attributes);
         } else {
             // * Not yet in use.
         }
Index: ggzd/ggzd/client.c
===================================================================
--- ggzd/ggzd/client.c	(revision 9253)
+++ ggzd/ggzd/client.c	(working copy)
@@ -41,6 +41,7 @@
 #include <string.h>
 #include <sys/socket.h>
 
+#include "admin.h"
 #include "client.h"
 #include "err_func.h"
 #include "hash.h"
@@ -126,6 +127,16 @@
 		client_free(client);
 		pthread_exit(NULL);
 	}
+
+	/* Check whether ip is banned */
+	if(is_banned(client->addr)){
+		net_send_player_banned(client->net, opt.server_name);
+		log_msg(GGZ_LOG_NOTICE,
+			"BANNED - Rejected connection from %s",
+			client->addr);
+		client_free(client);
+		pthread_exit(NULL);
+	}
 	
 	pthread_rwlock_wrlock(&state.lock);
 	if (state.players == MAX_USERS) {
Index: ggzd/ggzd/control.c
===================================================================
--- ggzd/ggzd/control.c	(revision 9253)
+++ ggzd/ggzd/control.c	(working copy)
@@ -46,6 +46,7 @@
 #include <avahi-client/publish.h>
 #endif
 
+#include "admin.h"
 #include "client.h"
 #include "chat.h"
 #include "daemon.h"
@@ -136,6 +137,7 @@
 		err_msg_exit("*** Database initialization failed");
 
 	hash_initialize();
+	bans_initialize();
 }
 
 static void cleanup_data(void)
@@ -714,3 +716,5 @@
 
 	return 0;
 }
+
+
Index: ggzd/ggzd/players.c
===================================================================
--- ggzd/ggzd/players.c	(revision 9253)
+++ ggzd/ggzd/players.c	(working copy)
@@ -1529,6 +1529,20 @@
 		break;
 	case GGZ_ADMIN_BAN:
 		log_msg(GGZ_LOG_ADMIN, "ADMIN %s bans %s", player->name, target);
+		if (rcvr == NULL || rcvr == player) {
+			status = E_USR_LOOKUP;
+		} else {
+			bans_add(rcvr->client->addr);
+			/* FIXME 1: session_over should be bool, not char */
+			/* FIXME: We temporarily take away the wr-lock since player_enqueue needs it */
+			pthread_rwlock_unlock(&rcvr->lock);	
+			chat_player_enqueue(target, GGZ_CHAT_PERSONAL, player, "You have been banned");
+
+			/* Recover the rcvr player - a new lookup is needed */
+			rcvr = hash_player_lookup(target);
+			if (rcvr) rcvr->client->session_over = 1;
+			status = E_OK;
+		}
 		/* This is not implemented/used yet. */
 		break;
 	}
Index: ggzd/ggzd/Makefile.am
===================================================================
--- ggzd/ggzd/Makefile.am	(revision 9253)
+++ ggzd/ggzd/Makefile.am	(working copy)
@@ -11,6 +11,8 @@
 ggzdconf_DATA = ggzd.conf.example ggzd.motd username.policy
 
 ggzd_SOURCES = \
+	admin.c \
+	admin.h \
 	chat.c \
 	chat.h \
 	client.c \
Index: ggzd/ggzd/admin.c
===================================================================
--- ggzd/ggzd/admin.c	(revision 0)
+++ ggzd/ggzd/admin.c	(revision 0)
@@ -0,0 +1,114 @@
+/*
+ * File: admin.c
+ * Author: Roger Light
+ * Project: GGZ Server
+ * Date: 25/08/2007
+ * Desc: Functions for handling admin actions.
+ * $Id$
+ *
+ * Copyright (C) 2007 Roger Light.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
+ */
+
+#ifdef HAVE_CONFIG_H
+# include <config.h>		/* Site specific config */
+#endif
+
+/*#include <ctype.h>
+#include <stdlib.h>
+*/
+
+#include <pthread.h>
+#include <string.h>
+#include <ggz.h>
+
+#include "admin.h"
+#include "err_func.h"
+
+
+/* Internal variables */
+static BanList bans;
+
+
+void bans_initialize(void)
+{
+	pthread_rwlock_init(&bans.lock, NULL);
+	bans.list = ggz_list_create(NULL, NULL, NULL, GGZ_LIST_ALLOW_DUPS);
+}
+
+
+void bans_add(const char *addr)
+{
+	BanItem *ban;
+	time_t t;
+
+	if(!addr) return;
+
+	ban = ggz_malloc(sizeof(BanItem));
+	if(!ban) return;
+
+	strncpy(ban->addr, addr, 64);
+	t = time(NULL);
+	if(t == -1){
+		ggz_free(ban);
+		return;
+	}
+
+	ban->expires = t + 900; /* Current time plus 15 minutes */
+	
+	pthread_rwlock_wrlock(&bans.lock);	
+	ggz_list_insert(bans.list, ban);
+	pthread_rwlock_unlock(&bans.lock);	
+}
+
+
+int is_banned(const char *addr)
+{
+	GGZListEntry *entry, *next;
+	BanItem *ban;
+	int rc = 0;
+
+	/*
+	 * This is a wrlock in case a ban has expired and needs
+	 * removing. Some sort of locking is required and assuming
+	 * that only a rdlock is required here means four lock 
+	 * operations are required when a ban needs to be expired.
+	 */
+	pthread_rwlock_wrlock(&bans.lock);
+	entry = ggz_list_head(bans.list);
+	while(entry){
+		/* Grab next entry here in case current entry is expired. */
+		next = ggz_list_next(entry);
+		ban = ggz_list_get_data(entry);
+
+#ifdef DEBUG
+		dbg_msg(GGZ_DBG_CONNECTION, "Banned IP: %s (%s)", ban->addr, addr);
+#endif
+		if(!strncmp(ban->addr, addr, 64)){
+			if(ban->expires >= time(NULL)){
+				rc = 1;
+				break;
+			}else{
+				ggz_list_delete_entry(bans.list, entry);
+				ggz_free(ban);
+			}
+		}
+		entry = next;
+	}
+	pthread_rwlock_unlock(&bans.lock);
+
+	return rc;
+}
Index: ggzd/ggzd/net.h
===================================================================
--- ggzd/ggzd/net.h	(revision 9253)
+++ ggzd/ggzd/net.h	(working copy)
@@ -56,6 +56,7 @@
 /* Functions to send data to the client */
 GGZReturn net_send_serverid(GGZNetIO *net, const char *srv_name, bool use_tls);
 GGZReturn net_send_server_full(GGZNetIO *net, const char *srv_name);
+GGZReturn net_send_player_banned(GGZNetIO *net, const char *srv_name);
 GGZReturn net_send_login(GGZNetIO *net, GGZLoginType type,
 			 GGZClientReqError status, const char *password);
 GGZReturn net_send_motd(GGZNetIO *net);
Index: ggzd/ggzd/admin.h
===================================================================
--- ggzd/ggzd/admin.h	(revision 0)
+++ ggzd/ggzd/admin.h	(revision 0)
@@ -0,0 +1,44 @@
+/*
+ * File: admin.h
+ * Author: Roger Light
+ * Project: GGZ Server
+ * Date: 25/08/2007
+ * Desc: Functions for handling admin actions.
+ * $Id$
+ *
+ * Copyright (C) 2007 Roger Light.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
+ */
+
+#include <pthread.h>
+#include <ggz.h>
+
+/* Ban data */
+typedef struct BanList {
+	pthread_rwlock_t lock;
+	GGZList *list;
+} BanList;
+typedef struct BanItem {
+	char addr[64];
+	time_t expires;
+} BanItem;
+
+
+/* Exported functions */
+void bans_initialize(void);
+void bans_add(const char *addr);
+int is_banned(const char *addr);
+
