Index: kreversi/kreversiui.rc
===================================================================
--- kreversi/kreversiui.rc	(Revision 846579)
+++ kreversi/kreversiui.rc	(Arbeitskopie)
@@ -4,6 +4,7 @@
 
 <MenuBar>
   <Menu name="game">
+    <Action name="game_connect" />
     <Action name="game_seats" />
   </Menu>
   <Menu name="view"><text>&amp;View</text>
Index: kreversi/mainwindow.h
===================================================================
--- kreversi/mainwindow.h	(Revision 846579)
+++ kreversi/mainwindow.h	(Arbeitskopie)
@@ -52,6 +52,7 @@
     void slotShowMovesHistory(bool);
     void slotHighscores();
     void slotSeats();
+    void slotConnect();
 private:
     virtual void showEvent( QShowEvent* );
     void setupActions();
@@ -77,6 +78,7 @@
     QAction* m_hintAct;
     QAction* m_demoAct;
     QAction* m_seatsAct;
+    QAction* m_connectAct;
     KSelectAction* m_animSpeedAct;
     KToggleAction* m_coloredChipsAct;
 };
Index: kreversi/CMakeLists.txt
===================================================================
--- kreversi/CMakeLists.txt	(Revision 846579)
+++ kreversi/CMakeLists.txt	(Arbeitskopie)
@@ -23,7 +23,7 @@
 kde4_add_app_icon(kreversi_SRCS "${CMAKE_CURRENT_SOURCE_DIR}/icons/hi*-app-kreversi.png")
 kde4_add_executable(kreversi ${kreversi_SRCS})
 
-target_link_libraries(kreversi ${KDE4_KDEUI_LIBS} kdegames ${KGGZGAMES_LIBS})
+target_link_libraries(kreversi ${KDE4_KDEUI_LIBS} kdegames ${KGGZGAMES_LIBS} kggzcore kggzlib)
 install(TARGETS kreversi  ${INSTALL_TARGETS_DEFAULT_ARGS} )
 
 register_ggz_module(module.dsc)
Index: kreversi/mainwindow.cpp
===================================================================
--- kreversi/mainwindow.cpp	(Revision 846579)
+++ kreversi/mainwindow.cpp	(Arbeitskopie)
@@ -28,6 +28,9 @@
 
 #include <kggzgames/kggzseatsdialog.h>
 #include <kggzmod/module.h>
+#include <kggzlib/connectiondialog.h>
+#include <kggzlib/kggzcorelayer.h>
+#include <kggzcore/player.h>
 #include <kaction.h>
 #include <kactioncollection.h>
 #include <ktoggleaction.h>
@@ -121,6 +124,12 @@
     m_hintAct = KStandardGameAction::hint(m_scene, SLOT(slotHint()), actionCollection());
     m_demoAct = KStandardGameAction::demo(this, SLOT(slotToggleDemoMode()), actionCollection());
 
+    m_connectAct = actionCollection()->addAction( "game_connect" );
+    m_connectAct->setIcon( KIcon("network-connect") );
+    m_connectAct->setText( i18n("Connect to GGZ Gaming Zone") );
+    m_connectAct->setShortcut( Qt::Key_C );
+    connect(m_connectAct, SIGNAL(triggered(bool)), SLOT(slotConnect()) );
+
     m_seatsAct = actionCollection()->addAction( "game_seats" );
     m_seatsAct->setIcon( KIcon("roll") );
     m_seatsAct->setText( i18n("Players and Seats") );
@@ -426,6 +435,16 @@
     }
 }
 
+void KReversiMainWindow::slotConnect()
+{
+	ConnectionDialog dialog(this);
+	dialog.setGame("Reversi", "4");
+	if(dialog.exec() == QDialog::Accepted)
+	{
+		slotNewGame();
+	}
+}
+
 void KReversiMainWindow::slotSeats()
 {
     KGGZSeatsDialog *dlg = new KGGZSeatsDialog();
Index: ksquares/src/ksquaresui.rc
===================================================================
--- ksquares/src/ksquaresui.rc	(Revision 846579)
+++ ksquares/src/ksquaresui.rc	(Arbeitskopie)
@@ -11,6 +11,7 @@
 	<MenuBar>
 	<Menu noMerge="1" name="onlinegame">
 		<text>Online game</text>
+		<Action name="connection" />
 		<Action name="rankings" />
 		<Action name="seats" />
 	</Menu>
Index: ksquares/src/ksquareswindow.h
===================================================================
--- ksquares/src/ksquareswindow.h	(Revision 846579)
+++ ksquares/src/ksquareswindow.h	(Arbeitskopie)
@@ -60,6 +60,7 @@
 		void slotMoveRequest(int x1, int y1, int x2, int y2);
 		void slotRankingsRequest();
 		void slotSeatsRequest();
+		void slotConnection();
 
 	private:
 		enum StatusBarItem
@@ -70,6 +71,7 @@
 
 		//void setupAccel();
 		void setupActions();
+		void initGGZ();
 		Ui::prefs_ai ui_prefs_ai;
 		Ui::prefs_display ui_prefs_display;
 		///The game board view
Index: ksquares/src/CMakeLists.txt
===================================================================
--- ksquares/src/CMakeLists.txt	(Revision 846579)
+++ ksquares/src/CMakeLists.txt	(Arbeitskopie)
@@ -22,7 +22,7 @@
 kde4_add_app_icon(ksquares_SRCS "${CMAKE_CURRENT_SOURCE_DIR}/hi*-app-ksquares.png")
 kde4_add_executable(ksquares ${ksquares_SRCS})
 
-target_link_libraries(ksquares ${KDE4_KDEUI_LIBS} kdegames ${KGGZGAMES_LIBS})
+target_link_libraries(ksquares ${KDE4_KDEUI_LIBS} kdegames ${KGGZGAMES_LIBS} kggzlib)
 
 install(TARGETS ksquares ${INSTALL_TARGETS_DEFAULT_ARGS})
 
Index: ksquares/src/ksquareswindow.cpp
===================================================================
--- ksquares/src/ksquareswindow.cpp	(Revision 846579)
+++ ksquares/src/ksquareswindow.cpp	(Arbeitskopie)
@@ -31,6 +31,9 @@
 #include <kggzmod/module.h>
 #include <kggzgames/kggzrankingsdialog.h>
 #include <kggzgames/kggzseatsdialog.h>
+#include <kggzlib/connectiondialog.h>
+#include <kggzlib/kggzcorelayer.h>
+#include <kggzcore/player.h>
 
 //generated
 #include "settings.h"
@@ -50,9 +53,7 @@
 
 	if(KGGZMod::Module::isGGZ())
 	{
-		KGGZMod::Module *mod = new KGGZMod::Module("ksquares");
-		connect(mod, SIGNAL(signalNetwork(int)), SLOT(slotNetworkData(int)));
-		connect(mod, SIGNAL(signalError()), SLOT(slotNetworkError()));
+		initGGZ();
 	}
 }
 
@@ -65,6 +66,13 @@
         delete m_proto;
 }
 
+void KSquaresWindow::initGGZ()
+{
+	KGGZMod::Module *mod = new KGGZMod::Module("ksquares");
+	connect(mod, SIGNAL(signalNetwork(int)), SLOT(slotNetworkData(int)));
+	connect(mod, SIGNAL(signalError()), SLOT(slotNetworkError()));
+}
+
 void KSquaresWindow::initObject()
 {
 	sGame = new KSquaresGame();
@@ -330,6 +338,10 @@
 	KStandardGameAction::gameNew(this, SLOT(gameNew()), actionCollection());
 	KAction *resetGame = KStandardGameAction::restart(this, SLOT(gameReset()), actionCollection());
 
+	KAction *a_connection = new KAction(this);
+	a_connection->setText(i18n("Connect to GGZ Gaming Zone"));
+	actionCollection()->addAction("connection", a_connection);
+
 	KAction *a_rankings = new KAction(this);
 	a_rankings->setText(i18n("Online rankings"));
 	actionCollection()->addAction("rankings", a_rankings);
@@ -345,6 +357,8 @@
 
 		KStandardGameAction::highscores(this, SLOT(showHighscores()), actionCollection());
 
+		connect(a_connection, SIGNAL(triggered(bool)), SLOT(slotConnection()));
+
 		a_rankings->setEnabled(false);
 		a_seats->setEnabled(false);
 	}
@@ -354,6 +368,8 @@
 
 		connect(a_rankings, SIGNAL(triggered(bool)), SLOT(slotRankingsRequest()));
 		connect(a_seats, SIGNAL(triggered(bool)), SLOT(slotSeatsRequest()));
+
+		a_connection->setEnabled(false);
 	}
 
 	KStandardGameAction::quit(this, SLOT(close()), actionCollection());
@@ -524,4 +540,14 @@
 	m_seatsdlg->show();
 }
 
+void KSquaresWindow::slotConnection()
+{
+	ConnectionDialog dialog(this);
+	dialog.setGame("Dots", "4");
+	if(dialog.exec() == QDialog::Accepted)
+	{
+		initGGZ();
+	}
+}
+
 #include "ksquareswindow.moc"
