Index: kreversiui.rc
===================================================================
--- kreversiui.rc	(Revision 839281)
+++ kreversiui.rc	(Arbeitskopie)
@@ -4,6 +4,7 @@
 
 <MenuBar>
   <Menu name="game">
+    <Action name="game_connect" />
     <Action name="game_seats" />
   </Menu>
   <Menu name="view"><text>&amp;View</text>
Index: mainwindow.h
===================================================================
--- mainwindow.h	(Revision 839281)
+++ mainwindow.h	(Arbeitskopie)
@@ -52,6 +52,7 @@
     void slotShowMovesHistory(bool);
     void slotHighscores();
     void slotSeats();
+    void slotConnect();
 private:
     virtual void showEvent( QShowEvent* );
     void setupActions();
@@ -77,6 +78,7 @@
     QAction* m_hintAct;
     QAction* m_demoAct;
     QAction* m_seatsAct;
+    QAction* m_connectAct;
     KSelectAction* m_animSpeedAct;
     KToggleAction* m_coloredChipsAct;
 };
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(Revision 839281)
+++ CMakeLists.txt	(Arbeitskopie)
@@ -23,7 +23,7 @@
 kde4_add_app_icon(kreversi_SRCS "${CMAKE_CURRENT_SOURCE_DIR}/icons/hi*-app-kreversi.png")
 kde4_add_executable(kreversi ${kreversi_SRCS})
 
-target_link_libraries(kreversi ${KDE4_KDEUI_LIBS} kdegames ${KGGZGAMES_LIBS})
+target_link_libraries(kreversi ${KDE4_KDEUI_LIBS} kdegames ${KGGZGAMES_LIBS} kggzcore kggzlib)
 install(TARGETS kreversi  ${INSTALL_TARGETS_DEFAULT_ARGS} )
 
 register_ggz_module(module.dsc)
Index: mainwindow.cpp
===================================================================
--- mainwindow.cpp	(Revision 839281)
+++ mainwindow.cpp	(Arbeitskopie)
@@ -28,6 +28,9 @@
 
 #include <kggzgames/kggzseatsdialog.h>
 #include <kggzmod/module.h>
+#include <kggzlib/connectiondialog.h>
+#include <kggzlib/kggzcorelayer.h>
+#include <kggzcore/player.h>
 #include <kaction.h>
 #include <kactioncollection.h>
 #include <ktoggleaction.h>
@@ -121,6 +124,12 @@
     m_hintAct = KStandardGameAction::hint(m_scene, SLOT(slotHint()), actionCollection());
     m_demoAct = KStandardGameAction::demo(this, SLOT(slotToggleDemoMode()), actionCollection());
 
+    m_connectAct = actionCollection()->addAction( "game_connect" );
+    m_connectAct->setIcon( KIcon("network-connect") );
+    m_connectAct->setText( i18n("Connect to GGZ Gaming Zone") );
+    m_connectAct->setShortcut( Qt::Key_C );
+    connect(m_connectAct, SIGNAL(triggered(bool)), SLOT(slotConnect()) );
+
     m_seatsAct = actionCollection()->addAction( "game_seats" );
     m_seatsAct->setIcon( KIcon("roll") );
     m_seatsAct->setText( i18n("Players and Seats") );
@@ -426,6 +435,20 @@
     }
 }
 
+void KReversiMainWindow::slotConnect()
+{
+	ConnectionDialog dialog(this);
+	dialog.setGame("Reversi", "4");
+	if(dialog.exec() == QDialog::Accepted)
+	{
+		QList<KGGZCore::Player> seats;
+		seats << KGGZCore::Player("somebot", KGGZCore::Player::bot);
+		dialog.layer()->configureTable(seats);
+
+		slotNewGame();
+	}
+}
+
 void KReversiMainWindow::slotSeats()
 {
     KGGZSeatsDialog *dlg = new KGGZSeatsDialog();
