diff -Nur teg-0.11.0.orig/client/ggz_client.c teg-0.11.0/client/ggz_client.c
--- teg-0.11.0.orig/client/ggz_client.c	2002-10-14 01:29:08.000000000 +0200
+++ teg-0.11.0/client/ggz_client.c	2004-06-13 00:06:20.000000000 +0200
@@ -53,7 +53,7 @@
 
 	g_ggzmod = ggzmod_new(GGZMOD_GAME);
 
-	ggzmod_set_handler( g_ggzmod, GGZMOD_EVENT_SERVER, &handle_ggzmod_server);
+	ggzmod_set_handler(g_ggzmod, GGZMOD_EVENT_SERVER, &handle_ggzmod_server);
 
 	g_is_initialized = 1;
 }
diff -Nur teg-0.11.0.orig/client/gui-gnome/connect.c teg-0.11.0/client/gui-gnome/connect.c
--- teg-0.11.0.orig/client/gui-gnome/connect.c	2002-10-14 23:04:36.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/connect.c	2004-06-13 00:04:42.000000000 +0200
@@ -39,6 +39,7 @@
 #include "connect.h"
 #include "priv.h"
 #include "colors.h"
+#include "ggz_client.h"
 
 static GtkWidget *connect_window = NULL;
 static GtkWidget *con_entry_name=NULL;
@@ -88,7 +89,7 @@
 		}
 #ifdef WITH_GGZ
 		else {
-			gui_private.tag_ggz = gdk_input_add( ggz_client_get_fd(), GDK_INPUT_READ, (GdkInputFunction) ggz_client_handle(), (gpointer) NULL );
+			gui_private.tag_ggz = gdk_input_add( ggz_client_get_fd(), GDK_INPUT_READ, (GdkInputFunction) ggz_client_handle, (gpointer) NULL );
 		}
 #endif /* WITH_GGZ */
 		countries_redraw_all();
diff -Nur teg-0.11.0.orig/client/gui-gnome/stock/stock-endturn.xpm teg-0.11.0/client/gui-gnome/stock/stock-endturn.xpm
--- teg-0.11.0.orig/client/gui-gnome/stock/stock-endturn.xpm	2001-11-02 04:00:46.000000000 +0100
+++ teg-0.11.0/client/gui-gnome/stock/stock-endturn.xpm	2004-06-13 00:01:51.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_endturn_xpm[] = {
+static const char * stock_endturn_xpm[] = {
 "24 24 150 2",
 "  	c None",
 ". 	c #010000",
diff -Nur teg-0.11.0.orig/client/gui-gnome/stock/stock-getcard.xpm teg-0.11.0/client/gui-gnome/stock/stock-getcard.xpm
--- teg-0.11.0.orig/client/gui-gnome/stock/stock-getcard.xpm	2000-08-07 02:37:12.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/stock/stock-getcard.xpm	2004-06-13 00:02:24.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_getcard_xpm[] = {
+static const char * stock_getcard_xpm[] = {
 "24 24 197 2",
 "  	c None",
 ". 	c #F1EAAB",
diff -Nur teg-0.11.0.orig/client/gui-gnome/stock/stock-sendarmies.xpm teg-0.11.0/client/gui-gnome/stock/stock-sendarmies.xpm
--- teg-0.11.0.orig/client/gui-gnome/stock/stock-sendarmies.xpm	2000-08-07 02:33:53.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/stock/stock-sendarmies.xpm	2004-06-13 00:02:13.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_sendarmies_xpm[] = {
+static const char * stock_sendarmies_xpm[] = {
 "24 24 9 1",
 " 	c None",
 ".	c #000000",
diff -Nur teg-0.11.0.orig/client/gui-gnome/stock/stock-viewplayers.xpm teg-0.11.0/client/gui-gnome/stock/stock-viewplayers.xpm
--- teg-0.11.0.orig/client/gui-gnome/stock/stock-viewplayers.xpm	2000-08-14 03:50:28.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/stock/stock-viewplayers.xpm	2004-06-13 00:02:03.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_viewplayers_xpm[] = {
+static const char * stock_viewplayers_xpm[] = {
 "24 24 9 1",
 " 	c None",
 ".	c #000000",
diff -Nur teg-0.11.0.orig/client/misc.c teg-0.11.0/client/misc.c
--- teg-0.11.0.orig/client/misc.c	2002-10-14 01:29:08.000000000 +0200
+++ teg-0.11.0/client/misc.c	2004-06-13 00:05:19.000000000 +0200
@@ -28,6 +28,7 @@
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <dirent.h>
+#include <string.h>
 
 #include <glib.h>
 
diff -Nur teg-0.11.0.orig/common/common.c teg-0.11.0/common/common.c
--- teg-0.11.0.orig/common/common.c	2002-10-14 05:26:36.000000000 +0200
+++ teg-0.11.0/common/common.c	2004-06-12 23:57:17.000000000 +0200
@@ -34,6 +34,7 @@
 #include <stdio.h>
 #include <unistd.h>
 #include <assert.h>
+#include <string.h>
 
 char *g_colores[] = {
 	N_("red"),
diff -Nur teg-0.11.0.orig/common/stats.c teg-0.11.0/common/stats.c
--- teg-0.11.0.orig/common/stats.c	2002-09-19 04:32:15.000000000 +0200
+++ teg-0.11.0/common/stats.c	2004-06-12 23:57:51.000000000 +0200
@@ -26,6 +26,8 @@
 #include "stats.h"
 #include "common.h"
 
+#include <string.h>
+
 TEG_STATUS stats_init( PPLAYER_STATS s )
 {
 	memset( s, 0, sizeof(*s) );
diff -Nur teg-0.11.0.orig/ggz/README teg-0.11.0/ggz/README
--- teg-0.11.0.orig/ggz/README	2002-04-07 02:26:01.000000000 +0200
+++ teg-0.11.0/ggz/README	2004-06-13 00:54:29.000000000 +0200
@@ -1,29 +1,32 @@
 Making TEG work with GGZ
 ------------------------
 
-since TEG v0.9.3 works with GGZ v0.0.5 (the actual version is 0.0.5pre)
-
-
-from TEG v0.6.0 until v0.9.2, TEG works with GGZ v0.0.4
-
-
 Download GGZ from:
-	http://ggz.sourceforge.net
+	http://www.ggzgamingzone.org/
 
+Supported features:
+	(x) game spectators
+	(x) statistics
+	(x) bots
 
 Installing TEG with GGZ
 -----------------------
 
-As always:
+First, you'll have to switch on ggz mode.
+This needs ggzmod, ggzdmod, ggz-config and ggzd installed.
 
    ./configure --enable-ggz=yes
+
+Then, install as always:
+
    make
    su -
    make install
 
+History
+-------
+
+since TEG 0.11+ works with GGZ v0.0.9 including statistics
+since TEG v0.9.3 works with GGZ v0.0.5 (the actual version is 0.0.5pre)
+from TEG v0.6.0 until v0.9.2, TEG works with GGZ v0.0.4
 
-   Then:
-    cd ggz
-    ggz-config --install --modfile=tegclient.dsc
-    cp teg.dsc /usr/local/etc/ggzd/games
-    cp teg.room /usr/local/etc/ggzd/rooms
diff -Nur teg-0.11.0.orig/ggz/tegclient.dsc.in teg-0.11.0/ggz/tegclient.dsc.in
--- teg-0.11.0.orig/ggz/tegclient.dsc.in	2002-04-09 03:18:26.000000000 +0200
+++ teg-0.11.0/ggz/tegclient.dsc.in	2004-06-13 01:00:25.000000000 +0200
@@ -7,5 +7,5 @@
 Homepage = http://teg.sourceforge.net
 Name = teg
 ProtocolEngine = TEG
-ProtocolVersion = 6
-Version = 6
+ProtocolVersion = 7
+Version = @VERSION@
diff -Nur teg-0.11.0.orig/ggz/teg.dsc.in teg-0.11.0/ggz/teg.dsc.in
--- teg-0.11.0.orig/ggz/teg.dsc.in	2002-10-14 05:25:57.000000000 +0200
+++ teg-0.11.0/ggz/teg.dsc.in	2004-06-13 01:36:41.000000000 +0200
@@ -10,10 +10,14 @@
 
 [Protocol]
 Engine = TEG
-Version = 6
+Version = 7
 
 [TableOptions]
 AllowLeave = 1
 BotsAllowed =  1..6
 PlayersAllowed = 2..6
 AllowSpectators = 1
+
+[Statistics]
+Records = 1
+Highscore = 1
diff -Nur teg-0.11.0.orig/po/es.po teg-0.11.0/po/es.po
--- teg-0.11.0.orig/po/es.po	2002-10-14 06:41:07.000000000 +0200
+++ teg-0.11.0/po/es.po	2004-06-12 20:11:26.000000000 +0200
@@ -8,7 +8,7 @@
 "Last-Translator:  <>\n"
 "Language-Team: English <en@li.org>\n"
 "MIME-Version: 1.0\n"
-"Content-Type: text/plain; charset=iso-8859-1\n"
+"Content-Type: text/plain; charset=utf-8\n"
 "Content-Transfer-Encoding: 8bit\n"
 
 #: common/common.c:39
diff -Nur teg-0.11.0.orig/README teg-0.11.0/README
--- teg-0.11.0.orig/README	2001-03-26 14:08:26.000000000 +0200
+++ teg-0.11.0/README	2004-06-13 00:57:00.000000000 +0200
@@ -3,3 +3,9 @@
 strategy game. Some rules are different.
 
 http://teg.sourceforge.net
+
+It can be configured to support the GGZ Gaming Zone; see the GGZ
+directory.
+
+http://www.ggzgamingzone.org/games/teg/
+
diff -Nur teg-0.11.0.orig/README.GGZ teg-0.11.0/README.GGZ
--- teg-0.11.0.orig/README.GGZ	2002-04-07 15:50:08.000000000 +0200
+++ teg-0.11.0/README.GGZ	1970-01-01 01:00:00.000000000 +0100
@@ -1,46 +0,0 @@
-GGZ:
-----
-
-GGZ is GGZ GAMING ZONE
-http://ggz.sourceforge.net
-
-TEG, since version 0.6.0 (until v0.9.2) has support for GGZ v0.0.4
-TEG version 0.9.3 has support for GGZ v0.0.5
-
-By default, TEG is NOT compiled with ggz support, 
-You need to do this:
-	./configure --enable-ggz=yes
-
-`tegserver' and `tegclient' will be invoked by the ggz client
-and ggz server with `--ggz' command line option.
-
-So, the game will run in standard mode (standalone mode) by
-default. And if you want to run in ggz mode, just run the
-ggz client, that it will enable the ggz mode passing the
---ggz as a command line option.
-
-
-Installing the *.dsc files:
----------------------------
-
-In the ggz/ subdir, are two files:
-	tegclient.dsc	-> teg client dsc file
-	teg.dsc		-> teg server dsc file
-	teg.room	-> teg server room file
-
-You must have ggz installed and the run:
-	ggz-config --install --modfile=./ggz/tegclient.dsc	
-
-	cp ./ggz/teg.dsc `ggz-config -c`/ggzd/games
-	cp ./ggz/teg.room `ggz-config-c`/ggzd/rooms
-
-
-Running:
---------
-	You'll need to have ggz and teg installed.
-
-	run:
-		ggzd	(the ggz server)
-		ggz-gtk	(the ggz-gtk client)
-
-		log in, enter in the TEG room and launch teg.
diff -Nur teg-0.11.0.orig/server/ggz_server.c teg-0.11.0/server/ggz_server.c
--- teg-0.11.0.orig/server/ggz_server.c	2002-10-14 05:26:36.000000000 +0200
+++ teg-0.11.0/server/ggz_server.c	2004-07-16 18:35:29.000000000 +0200
@@ -22,7 +22,6 @@
  * supports for GGZ code
  */
 
-
 #include <stdlib.h>
 #include <stdio.h>
 #include <string.h>
@@ -34,10 +33,12 @@
 #include <errno.h>
 
 #include "server.h"
+#include "player.h"
 
 #ifdef WITH_GGZ
 
 #include <ggzdmod.h>
+#include <ggzdmod-ggz.h>
 
 static GGZdMod *g_ggzdmod = NULL;
 
@@ -63,7 +64,7 @@
 
 	/* Setup GGZ game module */
 	ggzdmod_set_handler(g_ggzdmod, GGZDMOD_EVENT_STATE, &_handle_ggz_state);
-        ggzdmod_set_handler(g_ggzdmod, GGZDMOD_EVENT_LOG, &_handle_ggz_log);
+	ggzdmod_set_handler(g_ggzdmod, GGZDMOD_EVENT_LOG, &_handle_ggz_log);
 	ggzdmod_set_handler(g_ggzdmod, GGZDMOD_EVENT_ERROR, &_handle_ggz_error);
 
 	ggzdmod_set_handler(g_ggzdmod, GGZDMOD_EVENT_JOIN, &_handle_ggz_join);
@@ -242,6 +243,39 @@
 	return TEG_STATUS_ERROR;
 }
 
+void ggz_server_gameover( int fd )
+{
+	int max, i;
+	GGZSeat seat;
+	GGZGameResult results[TEG_MAX_PLAYERS];
+	int scores[TEG_MAX_PLAYERS];
+	PLIST_ENTRY pL;
+	PSPLAYER pJ;
+
+	max = ggzdmod_get_num_seats( g_ggzdmod );
+
+	for(i=0;i<max;i++) {
+		seat = ggzdmod_get_seat( g_ggzdmod, i );
+		if( seat.fd == fd ) {
+			results[i] = GGZ_GAME_WIN;
+		} else {
+			results[i] = GGZ_GAME_LOSS;
+		}
+
+		scores[i] = 0;
+		pL = g_list_player.Flink;
+		while( !IsListEmpty( &g_list_player ) && (pL != &g_list_player) ) {
+			pJ = (PSPLAYER) pL;
+			if( pJ->fd == seat.fd ){
+				scores[i] = pJ->player_stats.score;
+			}
+			pL = LIST_NEXT(pL);
+		}
+	}
+
+	ggzdmod_report_game( g_ggzdmod, NULL, results, scores );
+}
+
 int ggz_server_main_loop()
 {
 	if( teg_ggz_init() == -1 )
diff -Nur teg-0.11.0.orig/server/ggz_server.h teg-0.11.0/server/ggz_server.h
--- teg-0.11.0.orig/server/ggz_server.h	2002-10-14 03:25:04.000000000 +0200
+++ teg-0.11.0/server/ggz_server.h	2004-07-16 17:27:55.000000000 +0200
@@ -27,5 +27,6 @@
 
 int ggz_server_main_loop();
 TEG_STATUS ggz_server_find_ggzname( int fd, char *n, int len );
+void ggz_server_gameover( int fd );
 
 #endif /* __TEGS_TEGGGZ_H */
diff -Nur teg-0.11.0.orig/server/main.c teg-0.11.0/server/main.c
--- teg-0.11.0.orig/server/main.c	2002-10-14 03:25:04.000000000 +0200
+++ teg-0.11.0/server/main.c	2004-06-13 00:08:13.000000000 +0200
@@ -32,6 +32,7 @@
 #include <unistd.h>
 #include <netinet/in.h>
 #include <errno.h>
+#include <string.h>
 
 #include "server.h"
 #include "scores.h"
diff -Nur teg-0.11.0.orig/server/metaserver.c teg-0.11.0/server/metaserver.c
--- teg-0.11.0.orig/server/metaserver.c	2002-10-08 06:35:57.000000000 +0200
+++ teg-0.11.0/server/metaserver.c	2004-06-13 00:12:32.000000000 +0200
@@ -62,9 +62,12 @@
 #define	NOPTIONS  (sizeof(meta_options)/sizeof(meta_options[0]))
 
 /* threaded funcions ! */
-void * thread_meta_on( char * comment )
+void * thread_meta_on( void * commentptr )
 {
 	int s;
+	char *comment;
+	
+	comment = (char*)commentptr;
 
 	if( ! g_just_once ) {
 		con_text_out_wop(M_INF,_("Adding %s:%d to metaserver %s:%d...\n"),
diff -Nur teg-0.11.0.orig/server/missions.c teg-0.11.0/server/missions.c
--- teg-0.11.0.orig/server/missions.c	2002-10-05 00:56:35.000000000 +0200
+++ teg-0.11.0/server/missions.c	2004-06-13 00:09:06.000000000 +0200
@@ -19,7 +19,10 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
  */
 /* checks that the secret missions are accomplished */
+
 #include <assert.h>
+#include <string.h>
+
 #include "server.h"
 
 /* Do player pJ accomplished his secret mission ? */
Binary files teg-0.11.0.orig/server/missions.o and teg-0.11.0/server/missions.o differ
Binary files teg-0.11.0.orig/server/options.o and teg-0.11.0/server/options.o differ
diff -Nur teg-0.11.0.orig/server/play.c teg-0.11.0/server/play.c
--- teg-0.11.0.orig/server/play.c	2002-10-14 03:25:04.000000000 +0200
+++ teg-0.11.0/server/play.c	2004-07-16 18:19:26.000000000 +0200
@@ -693,6 +693,9 @@
 
 		/* Did 'src' player win the game ? */
 		if( mission_chequear( pJ_src ) == TEG_STATUS_GAMEOVER || game_is_finished() ) {
+#ifdef WITH_GGZ
+			ggz_server_gameover(pJ_src->fd);
+#endif
 			con_text_out(M_INF,_("Player %s(%d) is the winner! Game Over\n"),pJ_src->name,pJ_src->numjug);
 			pJ_src->estado = PLAYER_STATUS_GAMEOVER;
 			game_end( pJ_src );
