diff -Nur buggy.teg-0.11.1/client/ggz_client.c teg-0.11.1/client/ggz_client.c
--- buggy.teg-0.11.1/client/ggz_client.c	2004-08-01 20:58:04.000000000 +0200
+++ teg-0.11.1/client/ggz_client.c	2004-08-01 20:00:24.000000000 +0200
@@ -72,7 +72,7 @@
 	while( g_fd == -1 ) {
 		ggzmod_dispatch( g_ggzmod );
 
-		if( ++i > 10 )
+		if( ++i > 10000000 )
 			break;
 	}
 
Binary files buggy.teg-0.11.1/client/libclient.a and teg-0.11.1/client/libclient.a differ
diff -Nur buggy.teg-0.11.1/ggz/teg.dsc teg-0.11.1/ggz/teg.dsc
--- buggy.teg-0.11.1/ggz/teg.dsc	2004-08-01 20:57:42.000000000 +0200
+++ teg-0.11.1/ggz/teg.dsc	2004-08-01 23:21:35.000000000 +0200
@@ -20,4 +20,4 @@
 
 [Statistics]
 Records = 1
-Highscore = 1
+Highscores = 1
Binary files buggy.teg-0.11.1/robot/tegrobot and teg-0.11.1/robot/tegrobot differ
diff -Nur buggy.teg-0.11.1/server/helper.c teg-0.11.1/server/helper.c
--- buggy.teg-0.11.1/server/helper.c	2004-08-01 20:57:42.000000000 +0200
+++ teg-0.11.1/server/helper.c	2004-08-01 22:17:40.000000000 +0200
@@ -554,9 +554,12 @@
 
 	// sockets[0] is the client side of the robot
 	// sockets[1] is the server side
-	int r = socketpair(AF_LOCAL, SOCK_STREAM, 0, sockets );
-	if( r != 0 )
-		return TEG_STATUS_ERROR;
+	sockets[0] = 0;
+	while(sockets[0] < 3) {
+		int r = socketpair(AF_LOCAL, SOCK_STREAM, 0, sockets );
+		if( r != 0 )
+			return TEG_STATUS_ERROR;
+	}
 
 	// launch the server connected
 	if ( (pid = fork()) < 0) {
Binary files buggy.teg-0.11.1/server/helper.o and teg-0.11.1/server/helper.o differ
Binary files buggy.teg-0.11.1/server/tegserver and teg-0.11.1/server/tegserver differ
