diff -ur origs/teg-0.11.0.orig/client/ggz_client.c teg-0.11.0/client/ggz_client.c
--- origs/teg-0.11.0.orig/client/ggz_client.c	2002-10-14 01:29:08.000000000 +0200
+++ teg-0.11.0/client/ggz_client.c	2004-06-13 00:06:20.000000000 +0200
@@ -53,7 +53,7 @@
 
 	g_ggzmod = ggzmod_new(GGZMOD_GAME);
 
-	ggzmod_set_handler( g_ggzmod, GGZMOD_EVENT_SERVER, &handle_ggzmod_server);
+	ggzmod_set_handler(g_ggzmod, GGZMOD_EVENT_SERVER, &handle_ggzmod_server);
 
 	g_is_initialized = 1;
 }
diff -ur origs/teg-0.11.0.orig/client/gui-gnome/connect.c teg-0.11.0/client/gui-gnome/connect.c
--- origs/teg-0.11.0.orig/client/gui-gnome/connect.c	2002-10-14 23:04:36.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/connect.c	2004-06-13 00:04:42.000000000 +0200
@@ -39,6 +39,7 @@
 #include "connect.h"
 #include "priv.h"
 #include "colors.h"
+#include "ggz_client.h"
 
 static GtkWidget *connect_window = NULL;
 static GtkWidget *con_entry_name=NULL;
@@ -88,7 +89,7 @@
 		}
 #ifdef WITH_GGZ
 		else {
-			gui_private.tag_ggz = gdk_input_add( ggz_client_get_fd(), GDK_INPUT_READ, (GdkInputFunction) ggz_client_handle(), (gpointer) NULL );
+			gui_private.tag_ggz = gdk_input_add( ggz_client_get_fd(), GDK_INPUT_READ, (GdkInputFunction) ggz_client_handle, (gpointer) NULL );
 		}
 #endif /* WITH_GGZ */
 		countries_redraw_all();
diff -ur origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-endturn.xpm teg-0.11.0/client/gui-gnome/stock/stock-endturn.xpm
--- origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-endturn.xpm	2001-11-02 04:00:46.000000000 +0100
+++ teg-0.11.0/client/gui-gnome/stock/stock-endturn.xpm	2004-06-13 00:01:51.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_endturn_xpm[] = {
+static const char * stock_endturn_xpm[] = {
 "24 24 150 2",
 "  	c None",
 ". 	c #010000",
diff -ur origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-getcard.xpm teg-0.11.0/client/gui-gnome/stock/stock-getcard.xpm
--- origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-getcard.xpm	2000-08-07 02:37:12.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/stock/stock-getcard.xpm	2004-06-13 00:02:24.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_getcard_xpm[] = {
+static const char * stock_getcard_xpm[] = {
 "24 24 197 2",
 "  	c None",
 ". 	c #F1EAAB",
diff -ur origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-sendarmies.xpm teg-0.11.0/client/gui-gnome/stock/stock-sendarmies.xpm
--- origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-sendarmies.xpm	2000-08-07 02:33:53.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/stock/stock-sendarmies.xpm	2004-06-13 00:02:13.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_sendarmies_xpm[] = {
+static const char * stock_sendarmies_xpm[] = {
 "24 24 9 1",
 " 	c None",
 ".	c #000000",
diff -ur origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-viewplayers.xpm teg-0.11.0/client/gui-gnome/stock/stock-viewplayers.xpm
--- origs/teg-0.11.0.orig/client/gui-gnome/stock/stock-viewplayers.xpm	2000-08-14 03:50:28.000000000 +0200
+++ teg-0.11.0/client/gui-gnome/stock/stock-viewplayers.xpm	2004-06-13 00:02:03.000000000 +0200
@@ -1,5 +1,5 @@
 /* XPM */
-static char * stock_viewplayers_xpm[] = {
+static const char * stock_viewplayers_xpm[] = {
 "24 24 9 1",
 " 	c None",
 ".	c #000000",
diff -ur origs/teg-0.11.0.orig/client/misc.c teg-0.11.0/client/misc.c
--- origs/teg-0.11.0.orig/client/misc.c	2002-10-14 01:29:08.000000000 +0200
+++ teg-0.11.0/client/misc.c	2004-06-13 00:05:19.000000000 +0200
@@ -28,6 +28,7 @@
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <dirent.h>
+#include <string.h>
 
 #include <glib.h>
 
diff -ur origs/teg-0.11.0.orig/common/common.c teg-0.11.0/common/common.c
--- origs/teg-0.11.0.orig/common/common.c	2002-10-14 05:26:36.000000000 +0200
+++ teg-0.11.0/common/common.c	2004-06-12 23:57:17.000000000 +0200
@@ -34,6 +34,7 @@
 #include <stdio.h>
 #include <unistd.h>
 #include <assert.h>
+#include <string.h>
 
 char *g_colores[] = {
 	N_("red"),
diff -ur origs/teg-0.11.0.orig/common/stats.c teg-0.11.0/common/stats.c
--- origs/teg-0.11.0.orig/common/stats.c	2002-09-19 04:32:15.000000000 +0200
+++ teg-0.11.0/common/stats.c	2004-06-12 23:57:51.000000000 +0200
@@ -26,6 +26,8 @@
 #include "stats.h"
 #include "common.h"
 
+#include <string.h>
+
 TEG_STATUS stats_init( PPLAYER_STATS s )
 {
 	memset( s, 0, sizeof(*s) );
diff -ur origs/teg-0.11.0.orig/server/ggz_server.c teg-0.11.0/server/ggz_server.c
--- origs/teg-0.11.0.orig/server/ggz_server.c	2002-10-14 05:26:36.000000000 +0200
+++ teg-0.11.0/server/ggz_server.c	2004-06-13 00:11:36.000000000 +0200
@@ -38,6 +38,7 @@
 #ifdef WITH_GGZ
 
 #include <ggzdmod.h>
+#include <ggzdmod-ggz.h>
 
 static GGZdMod *g_ggzdmod = NULL;
 
diff -ur origs/teg-0.11.0.orig/server/main.c teg-0.11.0/server/main.c
--- origs/teg-0.11.0.orig/server/main.c	2002-10-14 03:25:04.000000000 +0200
+++ teg-0.11.0/server/main.c	2004-06-13 00:08:13.000000000 +0200
@@ -32,6 +32,7 @@
 #include <unistd.h>
 #include <netinet/in.h>
 #include <errno.h>
+#include <string.h>
 
 #include "server.h"
 #include "scores.h"
diff -ur origs/teg-0.11.0.orig/server/metaserver.c teg-0.11.0/server/metaserver.c
--- origs/teg-0.11.0.orig/server/metaserver.c	2002-10-08 06:35:57.000000000 +0200
+++ teg-0.11.0/server/metaserver.c	2004-06-13 00:12:32.000000000 +0200
@@ -62,9 +62,12 @@
 #define	NOPTIONS  (sizeof(meta_options)/sizeof(meta_options[0]))
 
 /* threaded funcions ! */
-void * thread_meta_on( char * comment )
+void * thread_meta_on( void * commentptr )
 {
 	int s;
+	char *comment;
+	
+	comment = (char*)commentptr;
 
 	if( ! g_just_once ) {
 		con_text_out_wop(M_INF,_("Adding %s:%d to metaserver %s:%d...\n"),
diff -ur origs/teg-0.11.0.orig/server/missions.c teg-0.11.0/server/missions.c
--- origs/teg-0.11.0.orig/server/missions.c	2002-10-05 00:56:35.000000000 +0200
+++ teg-0.11.0/server/missions.c	2004-06-13 00:09:06.000000000 +0200
@@ -19,7 +19,10 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
  */
 /* checks that the secret missions are accomplished */
+
 #include <assert.h>
+#include <string.h>
+
 #include "server.h"
 
 /* Do player pJ accomplished his secret mission ? */
