Index: main.cc
===================================================================
RCS file: /cvsroot/widelands/widelands/src/main.cc,v
retrieving revision 1.54
diff -u -r1.54 main.cc
--- main.cc	26 Jan 2005 13:17:36 -0000	1.54
+++ main.cc	7 May 2005 16:04:33 -0000
@@ -285,7 +285,9 @@
             }
             break;
          }
-			else if(code == Fullscreen_Menu_NetSetup::GGZGAME) {
+			else if((code == Fullscreen_Menu_NetSetup::JOINGGZGAME)
+				|| (code == Fullscreen_Menu_NetSetup::HOSTGGZGAME)) {
+				if(code == Fullscreen_Menu_NetSetup::HOSTGGZGAME) NetGGZ::ref()->launch();
 				if(NetGGZ::ref()->host()) netgame = new NetHost();
 				else
 				{
Index: network_ggz.cc
===================================================================
RCS file: /cvsroot/widelands/widelands/src/network_ggz.cc,v
retrieving revision 1.1
diff -u -r1.1 network_ggz.cc
--- network_ggz.cc	29 Jan 2005 10:08:33 -0000	1.1
+++ network_ggz.cc	7 May 2005 16:04:33 -0000
@@ -93,7 +93,7 @@
 }
 
 #ifdef HAVE_GGZ
-void NetGGZ::ggzmod_server(GGZMod *mod, GGZModEvent e, void *data)
+void NetGGZ::ggzmod_server(GGZMod *mod, GGZModEvent e, const void *data)
 {
 	printf("GGZ ## ggzmod_server\n");
 	int fd = *(int*)data;
@@ -190,7 +190,7 @@
 	return ip_address;
 }
 
-void NetGGZ::initcore()
+void NetGGZ::initcore(const char *hostname)
 {
 #ifdef HAVE_GGZ
 	GGZOptions opt;
@@ -223,7 +223,7 @@
 	ggzcore_server_add_event_hook(ggzserver, GGZ_CHANNEL_CONNECTED, &NetGGZ::callback_server);
 	ggzcore_server_add_event_hook(ggzserver, GGZ_CHANNEL_READY, &NetGGZ::callback_server);
 
-	ggzcore_server_set_hostinfo(ggzserver, "live.ggzgamingzone.org", 5688, 0);
+	ggzcore_server_set_hostinfo(ggzserver, hostname, 5688, 0);
 	ggzcore_server_connect(ggzserver);
 
 	printf("GGZCORE ## start loop\n");
@@ -233,6 +233,17 @@
 #endif
 }
 
+void NetGGZ::deinitcore()
+{
+#ifdef HAVE_GGZ
+	ggzcore_server_logout(ggzserver);
+	ggzcore_server_disconnect(ggzserver);
+	ggzcore_server_free(ggzserver);
+	ggzserver = 0;
+	ggzcore_destroy();
+#endif
+}
+
 bool NetGGZ::usedcore()
 {
 #ifdef HAVE_GGZ
@@ -263,7 +274,7 @@
 }
 
 #ifdef HAVE_GGZ
-GGZHookReturn NetGGZ::callback_server(unsigned id, void *data, void *user)
+GGZHookReturn NetGGZ::callback_server(unsigned id, const void *data, const void *user)
 {
 	printf("GGZCORE ## callback: %i\n", id);
 	ggzobj->event_server(id, data);
@@ -273,7 +284,7 @@
 #endif
 
 #ifdef HAVE_GGZ
-GGZHookReturn NetGGZ::callback_room(unsigned id, void *data, void *user)
+GGZHookReturn NetGGZ::callback_room(unsigned id, const void *data, const void *user)
 {
 	printf("GGZCORE/room ## callback: %i\n", id);
 	ggzobj->event_room(id, data);
@@ -283,7 +294,7 @@
 #endif
 
 #ifdef HAVE_GGZ
-GGZHookReturn NetGGZ::callback_game(unsigned id, void *data, void *user)
+GGZHookReturn NetGGZ::callback_game(unsigned id, const void *data, const void *user)
 {
 	printf("GGZCORE/game ## callback: %i\n", id);
 	ggzobj->event_game(id, data);
@@ -292,7 +303,7 @@
 }
 #endif
 
-void NetGGZ::event_server(unsigned int id, void *data)
+void NetGGZ::event_server(unsigned int id, const void *data)
 {
 #ifdef HAVE_GGZ
 	GGZRoom *room;
@@ -308,7 +319,11 @@
 			break;
 		case GGZ_NEGOTIATED:
 			printf("GGZCORE ## -- negotiated\n");
-			ggzcore_server_set_logininfo(ggzserver, GGZ_LOGIN_GUEST, "widelands#ggz", "");
+#if GGZCORE_VERSION_MICRO < 11
+			ggzcore_server_set_logininfo(ggzserver, GGZ_LOGIN_GUEST, "widelands#ggz", NULL);
+#else
+			ggzcore_server_set_logininfo(ggzserver, GGZ_LOGIN_GUEST, "widelands#ggz", NULL, NULL);
+#endif
 			ggzcore_server_login(ggzserver);
 			break;
 		case GGZ_LOGGED_IN:
@@ -371,7 +386,7 @@
 #endif
 }
 
-void NetGGZ::event_room(unsigned int id, void *data)
+void NetGGZ::event_room(unsigned int id, const void *data)
 {
 #ifdef HAVE_GGZ
 	GGZRoom *room;
@@ -399,18 +414,30 @@
 #endif
 }
 
-void NetGGZ::event_game(unsigned int id, void *data)
+void NetGGZ::event_game(unsigned int id, const void *data)
 {
 #ifdef HAVE_GGZ
 	GGZRoom *room;
 	GGZGame *game;
+	GGZTable *table;
+	GGZGameType *gametype;
 
 	switch(id)
 	{
 		case GGZ_GAME_PLAYING:
 			printf("GGZCORE/game ## -- playing\n");
 			room = ggzcore_server_get_cur_room(ggzserver);
-			ggzcore_room_join_table(room, tableid, 0);
+			if(tableid == -1)
+			{
+				table = ggzcore_table_new();
+				gametype = ggzcore_room_get_gametype(room);
+				ggzcore_table_init(table, gametype, "test", ggzcore_gametype_get_max_players(gametype));
+				for(int i = 1; i < ggzcore_gametype_get_max_players(gametype); i++)
+					ggzcore_table_set_seat(table, i, GGZ_SEAT_OPEN, NULL);
+				ggzcore_room_launch_table(room, table);
+				ggzcore_table_free(table);
+			}
+			else ggzcore_room_join_table(room, tableid, 0);
 			break;
 		case GGZ_GAME_LAUNCHED:
 			printf("GGZCORE/game ## -- launched\n");
@@ -477,3 +504,25 @@
 #endif
 }
 
+void NetGGZ::launch()
+{
+#ifdef HAVE_GGZ
+	GGZGame *game;
+
+	printf("GGZCORE ## launch table\n");
+
+	tableid = -1;
+
+	game = ggzcore_game_new();
+	ggzcore_game_init(game, ggzserver, NULL);
+
+	ggzcore_game_add_event_hook(game, GGZ_GAME_LAUNCHED, &NetGGZ::callback_game);
+	ggzcore_game_add_event_hook(game, GGZ_GAME_LAUNCH_FAIL, &NetGGZ::callback_game);
+	ggzcore_game_add_event_hook(game, GGZ_GAME_NEGOTIATED, &NetGGZ::callback_game);
+	ggzcore_game_add_event_hook(game, GGZ_GAME_NEGOTIATE_FAIL, &NetGGZ::callback_game);
+	ggzcore_game_add_event_hook(game, GGZ_GAME_PLAYING, &NetGGZ::callback_game);
+
+	ggzcore_game_launch(game);
+#endif
+}
+
Index: network_ggz.h
===================================================================
RCS file: /cvsroot/widelands/widelands/src/network_ggz.h,v
retrieving revision 1.1
diff -u -r1.1 network_ggz.h
--- network_ggz.h	29 Jan 2005 10:08:33 -0000	1.1
+++ network_ggz.h	7 May 2005 16:04:33 -0000
@@ -57,21 +57,23 @@
 		op_broadcast_ip = 4
 	};
 
-	void initcore();
+ 	void initcore(const char *hostname);
+ 	void deinitcore();
 	bool usedcore();
 	void datacore();
+ 	void launch();
 	void join(const char *tablename);
 
     private:
 #ifdef HAVE_GGZ
-	static void ggzmod_server(GGZMod *mod, GGZModEvent e, void *data);
-	static GGZHookReturn callback_server(unsigned int id, void *data, void *user);
-	static GGZHookReturn callback_room(unsigned int id, void *data, void *user);
-	static GGZHookReturn callback_game(unsigned int id, void *data, void *user);
+	static void ggzmod_server(GGZMod *mod, GGZModEvent e, const void *data);
+	static GGZHookReturn callback_server(unsigned int id, const void *data, const void *user);
+	static GGZHookReturn callback_room(unsigned int id, const void *data, const void *user);
+	static GGZHookReturn callback_game(unsigned int id, const void *data, const void *user);
 #endif
-	void event_server(unsigned int id, void *data);
-	void event_room(unsigned int id, void *data);
-	void event_game(unsigned int id, void *data);
+	void event_server(unsigned int id, const void *data);
+	void event_room(unsigned int id, const void *data);
+	void event_game(unsigned int id, const void *data);
 
 	bool use_ggz;
 	int fd;
Index: network_lan_promotion.cc
===================================================================
RCS file: /cvsroot/widelands/widelands/src/network_lan_promotion.cc,v
retrieving revision 1.4
diff -u -r1.4 network_lan_promotion.cc
--- network_lan_promotion.cc	1 Jan 2005 17:16:59 -0000	1.4
+++ network_lan_promotion.cc	7 May 2005 16:04:34 -0000
@@ -181,11 +181,18 @@
 
 LAN_Game_Finder::LAN_Game_Finder ()
 {
-    char magic[8];
-    
     callback=0;
     
     bind (WIDELANDS_LAN_DISCOVERY_PORT);
+
+    reset();
+}
+
+void LAN_Game_Finder::reset ()
+{
+    char magic[8];
+
+    opengames.clear();
     
     strncpy (magic, "QUERY", 8);
     magic[6]=LAN_PROMOTION_PROTOCOL_VERSION;
Index: network_lan_promotion.h
===================================================================
RCS file: /cvsroot/widelands/widelands/src/network_lan_promotion.h,v
retrieving revision 1.2
diff -u -r1.2 network_lan_promotion.h
--- network_lan_promotion.h	31 Dec 2004 17:34:48 -0000	1.2
+++ network_lan_promotion.h	7 May 2005 16:04:34 -0000
@@ -89,6 +89,7 @@
 	LAN_Game_Finder ();
 	~LAN_Game_Finder ();
 	
+	void reset ();
 	void run ();
 	
 	void set_callback (void(*)(int, const LAN_Open_Game*, void*), void*);
Index: options.cc
===================================================================
RCS file: /cvsroot/widelands/widelands/src/options.cc,v
retrieving revision 1.29
diff -u -r1.29 options.cc
--- options.cc	26 Jan 2005 13:17:36 -0000	1.29
+++ options.cc	7 May 2005 16:04:34 -0000
@@ -62,7 +62,6 @@
 " --coredump       Generates a core dump on segfaults instead of using the SDL\n"
 "\n"
 " --ggz            Starts game as GGZ Gaming Zone client (don't use!)\n"
-" --ggzcore        GGZ embedded mode (TEMPORARY -> make menu button!)\n"
 "\n"
 #ifdef DEBUG
 " --double         Start the game twice (for localhost network testing)\n\n"
@@ -114,10 +113,6 @@
 			NetGGZ::ref()->init();
 			continue;
 		}
-		if (!strcmp(opt, "ggzcore")) {
-			NetGGZ::ref()->initcore();
-			continue;
-		}
 
 #ifdef DEBUG
 		if (!strcmp(opt, "double")) {
Index: ui/ui_fs_menus/fullscreen_menu_netsetup.cc
===================================================================
RCS file: /cvsroot/widelands/widelands/src/ui/ui_fs_menus/fullscreen_menu_netsetup.cc,v
retrieving revision 1.10
diff -u -r1.10 fullscreen_menu_netsetup.cc
--- ui/ui_fs_menus/fullscreen_menu_netsetup.cc	26 Jan 2005 13:17:36 -0000	1.10
+++ ui/ui_fs_menus/fullscreen_menu_netsetup.cc	7 May 2005 16:04:34 -0000
@@ -31,11 +31,8 @@
 Fullscreen_Menu_NetSetup::Fullscreen_Menu_NetSetup ()
 	:Fullscreen_Menu_Base("singleplmenu.jpg") // change this
 {
-	if(!NetGGZ::ref()->usedcore())
-	{
-		discovery=new LAN_Game_Finder();
-		discovery->set_callback (discovery_callback, this);
-	}
+	discovery=new LAN_Game_Finder();
+	discovery->set_callback (discovery_callback, this);
 	
 	// Text
 	UITextarea* title= new UITextarea(this, MENU_XRES/2, 140, "Begin Network Game", Align_HCenter);
@@ -45,11 +42,11 @@
 	UIButton* b;
 
 	b = new UIButton(this, 60, 170, 174, 24, 1, JOINGAME);
-	b->clickedid.set(this, &Fullscreen_Menu_NetSetup::end_modal);
+	b->clickedid.set(this, &Fullscreen_Menu_NetSetup::joingame);
 	b->set_title("Join a Game");
 
 	b = new UIButton(this, 60, 210, 174, 24, 1, HOSTGAME);
-	b->clickedid.set(this, &Fullscreen_Menu_NetSetup::end_modal);
+	b->clickedid.set(this, &Fullscreen_Menu_NetSetup::hostgame);
 	b->set_title("Host a New Game");
 
 	b = new UIButton(this, 60, 250, 174, 24, 1, INTERNETGAME);
@@ -59,10 +56,16 @@
 	b = new UIButton(this, 60, 290, 174, 24, 0, CANCEL);
 	b->clickedid.set(this, &Fullscreen_Menu_NetSetup::end_modal);
 	b->set_title("Back");
-	
+
 	// Hostname
 	hostname=new UIEdit_Box(this, 288, 170, 174, 24, 2, 0);
 	hostname->set_text("localhost");	
+
+	// LAN or GGZ game
+	networktype = new UIButton(this, 482, 170, 124, 24, 0, -1);
+	networktype->clickedid.set(this, &Fullscreen_Menu_NetSetup::toggle_networktype);
+	networktype->set_title("LAN games");
+	internetgame = false;
 	
 	// List of open games in local network
 	opengames=new UITable(this, 288, 210, 320, 128);
@@ -74,17 +77,17 @@
 
 Fullscreen_Menu_NetSetup::~Fullscreen_Menu_NetSetup ()
 {
-	if(!NetGGZ::ref()->usedcore())
-	{
-		delete discovery;
-	}
+	delete discovery;
 }
 
 void Fullscreen_Menu_NetSetup::think ()
 {
 	Fullscreen_Menu_Base::think ();
 	
-	discovery->run ();
+	if(!NetGGZ::ref()->usedcore())
+	{
+		discovery->run ();
+	}
 }
 
 bool Fullscreen_Menu_NetSetup::get_host_address (ulong& addr, ushort& port)
@@ -116,13 +119,7 @@
 {
 	LAN_Open_Game* game=(LAN_Open_Game*) (opengames->get_selection());
 	
-	if(game)
-		hostname->set_text (game->info.hostname);
-	else {
-		UITable_Entry *entry = opengames->get_entry(opengames->get_selection_index());
-		NetGGZ::ref()->join(entry->get_string(1));
-		end_modal(GGZGAME);
-	}
+	if(game) hostname->set_text (game->info.hostname);
 }
 
 void Fullscreen_Menu_NetSetup::update_game_info (UITable_Entry* entry, const LAN_Game_Info& info)
@@ -190,3 +187,53 @@
 	}
 }
 
+void Fullscreen_Menu_NetSetup::toggle_networktype(int code)
+{
+	if(internetgame)
+	{
+		NetGGZ::ref()->deinitcore();
+	}
+
+	internetgame = !internetgame;
+
+	opengames->clear();
+
+	if(internetgame)
+	{
+		hostname->set_text("live.ggzgamingzone.org");
+		NetGGZ::ref()->initcore(hostname->get_text());
+		networktype->set_title("GGZ games");
+	}
+	else
+	{
+		hostname->set_text("localhost");
+		discovery->reset();
+		networktype->set_title("LAN games");
+	}
+}
+
+//bool Fullscreen_Menu_NetSetup::is_internetgame()
+//{
+//	return internetgame;
+//}
+
+void Fullscreen_Menu_NetSetup::joingame(int code)
+{
+	if(NetGGZ::ref()->usedcore())
+	{
+		int index = opengames->get_selection_index();
+		if(index < 0) return;
+		UITable_Entry *entry = opengames->get_entry(index);
+		if(!entry) return;
+		NetGGZ::ref()->join(entry->get_string(1));
+		end_modal(JOINGGZGAME);
+	}
+	else end_modal(JOINGAME);
+}
+
+void Fullscreen_Menu_NetSetup::hostgame(int code)
+{
+	if(NetGGZ::ref()->usedcore()) end_modal(HOSTGGZGAME);
+	else end_modal(HOSTGAME);
+}
+
Index: ui/ui_fs_menus/fullscreen_menu_netsetup.h
===================================================================
RCS file: /cvsroot/widelands/widelands/src/ui/ui_fs_menus/fullscreen_menu_netsetup.h,v
retrieving revision 1.7
diff -u -r1.7 fullscreen_menu_netsetup.h
--- ui/ui_fs_menus/fullscreen_menu_netsetup.h	12 Jan 2005 01:30:54 -0000	1.7
+++ ui/ui_fs_menus/fullscreen_menu_netsetup.h	7 May 2005 16:04:34 -0000
@@ -28,6 +28,7 @@
 class UIEdit_Box;
 class UITable;
 class UITable_Entry;
+class UIButton;
 
 class LAN_Game_Finder;
 class LAN_Open_Game;
@@ -40,7 +41,8 @@
 			HOSTGAME,
 			JOINGAME,
 			INTERNETGAME,
-			GGZGAME
+			HOSTGGZGAME,
+			JOINGGZGAME
 		};
 		
 		Fullscreen_Menu_NetSetup ();
@@ -51,12 +53,17 @@
 		bool get_host_address (ulong&, ushort&);
 		// return true if the selected or entered hostname is valid
 
+		//bool is_internetgame();
+		// return true if game should be played over GGZ
+
 		void fill(std::list<std::string> tables);
 	
 	private:
 		UIEdit_Box*		hostname;
 		UITable*		opengames;
 		LAN_Game_Finder*	discovery;
+		UIButton*		networktype;
+		bool			internetgame;
 		
 		void game_selected (int);
 		
@@ -67,6 +74,10 @@
 		void game_updated (const LAN_Open_Game*);
 		
 		void update_game_info (UITable_Entry*, const LAN_Game_Info&);
+
+		void toggle_networktype(int code);
+		void joingame(int code);
+		void hostgame(int code);
 };
 
 #endif
