Index: ggzd/ggzd/players.c
===================================================================
--- ggzd/ggzd/players.c	(Revision 9944)
+++ ggzd/ggzd/players.c	(Arbeitskopie)
@@ -382,10 +382,10 @@
 		dbg_msg(GGZ_DBG_TABLE,
 			"%s tries to launch table without connection",
 			player->name);
-		if (net_send_table_launch(player->client->net,
-					  E_NO_CHANNEL) < 0)
-			return GGZ_REQ_DISCONNECT;
-		return GGZ_REQ_FAIL;
+//		if (net_send_table_launch(player->client->net,
+//					  E_NO_CHANNEL) < 0)
+//			return GGZ_REQ_DISCONNECT;
+//		return GGZ_REQ_FAIL;
 	}
 
 	/* Do actual launch of table */
Index: ggz-client-libs/ggzcore/room.c
===================================================================
--- ggz-client-libs/ggzcore/room.c	(Revision 9944)
+++ ggz-client-libs/ggzcore/room.c	(Arbeitskopie)
@@ -1078,8 +1078,10 @@
 	/* Make sure we're actually in a room (FIXME: should probably
 	   make sure we're in *this* room) and not already playing a
 	   game */
-	if (_ggzcore_server_get_state(room->server) != GGZ_STATE_IN_ROOM
-	    || !game)
+//	if (_ggzcore_server_get_state(room->server) != GGZ_STATE_IN_ROOM
+//	    || !game)
+//		return -1;
+	if (_ggzcore_server_get_state(room->server) != GGZ_STATE_IN_ROOM)
 		return -1;
 
 	net = _ggzcore_server_get_net(room->server);
