diff -Nur ggzd/acconfig.h ggzd.tls/acconfig.h
--- ggzd/acconfig.h	Fri Nov 10 08:36:04 2000
+++ ggzd.tls/acconfig.h	Mon Nov  5 23:24:55 2001
@@ -28,6 +28,7 @@
 #undef _XOPEN_SOURCE
 #undef DB2_IN_DIR
 #undef DB3_IN_DIR
+#undef GENSECURE
     
 
 /* Leave that blank line there!!  Autoheader needs it.
diff -Nur ggzd/configure.in ggzd.tls/configure.in
--- ggzd/configure.in	Mon Nov  5 00:30:26 2001
+++ ggzd.tls/configure.in	Tue Nov  6 00:09:04 2001
@@ -104,6 +104,14 @@
         AC_MSG_ERROR(*** Cannot find expat.h header)),
     AC_MSG_ERROR(*** Cannot find expat library ))
 
+dnl Check for gensecure
+dnl ===================
+AC_CHECK_LIB(gensecure, tls_prepare,
+	AC_CHECK_HEADER(gensecure.h,
+		INCLUDES="$INCLUDES -I/usr/local/ssl/include" AC_DEFINE_UNQUOTED(GENSECURE, "1") ggzd_LIBS="$ggzd_LIBS -lgensecure",
+		AC_MSG_ERROR(*** Cannot find gensecure.h header)),
+	AC_MSG_ERROR(*** Cannot find libgensecure ))
+
  
 dnl Check database libs
 dnl ===================
diff -Nur ggzd/ggzd/control.c ggzd.tls/ggzd/control.c
--- ggzd/ggzd/control.c	Fri Oct  5 01:44:50 2001
+++ ggzd.tls/ggzd/control.c	Tue Nov  6 00:17:03 2001
@@ -54,6 +54,7 @@
 #include <table.h>
 #include <daemon.h>
 #include <hash.h>
+#include <gensecure.h>
 
 
 /* Server options */
@@ -147,6 +148,12 @@
 	srandom((unsigned)time(NULL));
 }
 
+/* The TLS password handler */
+int passwordcallback(char *buf, int size, int rwflag, void *userdata)
+{
+	strncpy(buf, opt.tls_password, size);
+	return strlen(opt.tls_password);
+}
 
 int main(int argc, const char *argv[])
 {
@@ -163,6 +170,7 @@
 	dbg_msg(GGZ_DBG_CONFIGURATION, "Conf file: %s", opt.local_conf);
 	dbg_msg(GGZ_DBG_CONFIGURATION, "Log level: %0X", log_info.log_types);
 	dbg_msg(GGZ_DBG_CONFIGURATION, "Main Port: %d", opt.main_port);
+	dbg_msg(GGZ_DBG_CONFIGURATION, "Encryption in use: %d", opt.tls_use);
 	
 	init_dirs();
 	init_data();
@@ -183,6 +191,9 @@
 	}
 	
 	signal(SIGPIPE, SIG_IGN);
+
+	/* Setup TLS callback */
+	tls_prepare(opt.tls_cert, opt.tls_key, passwordcallback);
 
 	/* Create SERVER socket on main_port */
 	main_sock = es_make_socket_or_die(ES_SERVER, opt.main_port, NULL);
diff -Nur ggzd/ggzd/datatypes.h ggzd.tls/ggzd/datatypes.h
--- ggzd/ggzd/datatypes.h	Mon Oct 29 00:36:54 2001
+++ ggzd.tls/ggzd/datatypes.h	Tue Nov  6 00:33:56 2001
@@ -52,6 +52,11 @@
 	int perform_lookups;
 	int ping_freq;
 	int lag_class[4];
+	int tls_use;
+	char *tls_key;
+	char *tls_cert;
+	char *tls_password;
+	char *servername;
 } Options;
 
 
diff -Nur ggzd/ggzd/ggzd.conf.in ggzd.tls/ggzd/ggzd.conf.in
--- ggzd/ggzd/ggzd.conf.in	Mon Oct 29 00:36:54 2001
+++ ggzd.tls/ggzd/ggzd.conf.in	Mon Nov  5 23:19:07 2001
@@ -47,6 +47,12 @@
 #ignore = 167.14.108.115/255.255.255.192
 #  (ignores all addresses from 167.14.108.64 - 167.14.108.127)
 
+# Transport Layer Security options
+EncryptionUse = 1
+EncryptionPassword = ggz rocks more
+EncryptionKey = /usr/local/ssl/certs/openssl-example.pem
+EncryptionCert = /usr/local/ssl/certs/openssl-example.pem
+
 ############################################################################
 # Logging Options
 ############################################################################
diff -Nur ggzd/ggzd/net.c ggzd.tls/ggzd/net.c
--- ggzd/ggzd/net.c	Sat Nov  3 02:38:19 2001
+++ ggzd.tls/ggzd/net.c	Tue Nov  6 00:38:17 2001
@@ -43,6 +43,7 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <unistd.h>
+#include <gensecure.h>
 
 /* For convenience */
 #define BUFFSIZE 8192
@@ -161,6 +162,8 @@
 	net->parsing = 0;
 	net->done = 0;
 
+	tls_start(fd, TLS_SERVER, TLS_NOVERIFY);
+
         /* Init parser */
         if (!(net->parser = XML_ParserCreate(NULL)))
                 err_sys_exit("Couldn't allocate memory for XML parser");
@@ -243,10 +246,10 @@
 
 /* net_send_XXX() functions for sending messages to the client */
 
-int net_send_serverid(GGZNetIO *net)
+int net_send_serverid(GGZNetIO *net, const char *servername)
 {
 	_net_send_line(net, "<SESSION>");
-	_net_send_line(net, "\t<SERVER ID='GGZ-%s' NAME='%s' VERSION='%d' STATUS='%s'>", VERSION, "Brent&apos;s Server", GGZ_CS_PROTO_VERSION, "ok");
+	_net_send_line(net, "\t<SERVER ID='GGZ-%s' NAME='%s' VERSION='%d' STATUS='%s'>", VERSION, servername, GGZ_CS_PROTO_VERSION, "ok");
 	_net_send_line(net, "\t\t<OPTIONS CHATLEN='%d'/>", MAX_CHAT_LEN);
 	_net_send_line(net, "\t</SERVER>");
 			
@@ -662,7 +665,7 @@
 	buf[0] = '\0';
 	for (i = 0; i < size; i++) {
 		sprintf(buf, "%d ", data[i]);
-		write(net->fd, buf, strlen(buf));
+		tls_write(net->fd, buf, strlen(buf));
 	}
 	_net_send_string(net, "]]></DATA>");
 
@@ -1517,7 +1520,7 @@
 	vsprintf(buf, line, ap);
 	va_end(ap);
 	strcat(buf, "\n");
-	return write(net->fd, buf, strlen(buf));
+	return tls_write(net->fd, buf, strlen(buf));
 }
 
 
@@ -1529,5 +1532,5 @@
 	va_start(ap, fmt);
 	vsprintf(buf, fmt, ap);
 	va_end(ap);
-	return write(net->fd, buf, strlen(buf));
+	return tls_write(net->fd, buf, strlen(buf));
 }
diff -Nur ggzd/ggzd/net.h ggzd.tls/ggzd/net.h
--- ggzd/ggzd/net.h	Tue Oct  9 02:04:59 2001
+++ ggzd.tls/ggzd/net.h	Tue Nov  6 00:38:27 2001
@@ -53,7 +53,7 @@
 void net_free(GGZNetIO* net);
 
 /* Functions to send data to the client */
-int net_send_serverid(GGZNetIO *net);
+int net_send_serverid(GGZNetIO *net, const char *servername);
 int net_send_server_full(GGZNetIO *net);
 int net_send_login(GGZNetIO *net, GGZLoginType type, char status, char *password);
 int net_send_motd(GGZNetIO *net);
diff -Nur ggzd/ggzd/parse_opt.c ggzd.tls/ggzd/parse_opt.c
--- ggzd/ggzd/parse_opt.c	Mon Oct 29 00:36:54 2001
+++ ggzd.tls/ggzd/parse_opt.c	Tue Nov  6 00:36:23 2001
@@ -253,6 +253,13 @@
 		opt.main_port = conf_read_int(ch, "General", "Port", 5688);
 	opt.admin_name = conf_read_string(ch, "General", "AdminName", NULL);
 	opt.admin_email = conf_read_string(ch, "General", "AdminEmail", NULL);
+	opt.servername = conf_read_string(ch, "General", "Server", "Anonymous server");
+
+	/* Enctyption in [General] */
+	opt.tls_use = conf_read_int(ch, "General", "EncryptionUse", 0);
+	opt.tls_password = conf_read_string(ch, "General", "EncryptionPassword", NULL);
+	opt.tls_cert = conf_read_string(ch, "General", "EncryptionCert", NULL);
+	opt.tls_key = conf_read_string(ch, "General", "EncryptionKey", NULL);
 
 	/* [Directories] */
 	opt.game_dir = conf_read_string(ch, "Directories", "GameDir", NULL);
diff -Nur ggzd/ggzd/players.c ggzd.tls/ggzd/players.c
--- ggzd/ggzd/players.c	Mon Oct 29 05:25:16 2001
+++ ggzd.tls/ggzd/players.c	Tue Nov  6 00:38:01 2001
@@ -167,7 +167,7 @@
 	player->next_ping = 0;			/* Don't ping until login */
 
 	/* Send server ID */
-	if (net_send_serverid(player->net) < 0)
+	if (net_send_serverid(player->net, opt.servername) < 0)
 		pthread_exit(NULL);
 	
 	pthread_rwlock_wrlock(&state.lock);
