diff -ur civil.orig/src/civil.py civil/src/civil.py
--- civil.orig/src/civil.py	Sat Sep 22 17:31:47 2001
+++ civil/src/civil.py	Sat Sep 22 18:11:18 2001
@@ -23,6 +23,7 @@
 from dialog          import *
 from sdl             import SDL
 from audio_manager   import AudioManager
+from net.connection  import Connection
 
 from playfield.playfield          import Playfield
 from playfield.terrain_layer      import TerrainLayer
@@ -68,6 +69,11 @@
         elif parameter in ('-u', '--union' ):
             # yep, setup hardcoded data
             pass
+
+		# check whether game was launched over ggz
+        elif parameter in ('-g', '--ggz'):
+            # ggz mode is on
+            properties.network_ggz = 1
         
 
 ###############################################################################################
@@ -136,8 +142,12 @@
     first visible menu and lets the player choose scenario etc. When the dialog returns with an
     ACCEPTED state the game is ready to continue to the normal event loop."""
 
-    # create a new maindialog and run it
-    if SetupNetwork ().run () != ACCEPTED:
+	# Either connect to ggz...
+    if properties.network_ggz:
+         scenario.connection = Connection (3, "ggz", 5688) 
+
+    # ...or create a new maindialog and run it
+    elif SetupNetwork ().run () != ACCEPTED:
         # not ok, so just go away
         pygame.quit () 
         sys.exit ( 0 )
diff -ur civil.orig/src/properties.py civil/src/properties.py
--- civil.orig/src/properties.py	Sat Sep 22 17:31:48 2001
+++ civil/src/properties.py	Sat Sep 22 17:59:37 2001
@@ -163,6 +163,7 @@
 ###############################################################################################
 # network settings
 network_port                 = 20000
+network_ggz                  = 0
 
 ###############################################################################################
 # unit settings
